src/scheduler/MpiWorld.cpp

```diff
@@ -912,53 +918,138 @@ void MpiWorld::gather(int sendRank,
+
+                recv(it.second.front(),
                      recvRank,
-                     recvChunk,
+                     rankData.get(),
                      recvType,
-                     recvCount,
+                     recvCount * it.second.size(),
                      nullptr,
                      faabric::MPIMessage::GATHER);
+
+                // Copy each received chunk to its offset
+                for (int r = 0; r < it.second.size(); r++) {
+                    memcpy(recvBuffer + (it.second.at(r) * recvOffset),
+                           rankData.get() + (r * recvOffset),
+                           recvOffset);
+                }
+            }
+        }
+    } else if (sendRank == localLeader) {
+        if (getHostForRank(recvRank) != thisHost) {
+            // If we are the local leader (but not the receiver of the gather)
+            // and the receiver is not co-located with us, we gather all our
+            // local rank's data and send it in a single message
+            auto rankData = std::make_unique<uint8_t[]>(
+              ranksForHost[thisHost].size() * sendOffset);
+
+            for (int r = 0; r < ranksForHost[thisHost].size(); r++) {
+                if (ranksForHost[thisHost].at(r) == sendRank) {
+                    // If gathering in-place from a rank that is not-receiving
+                    // means that the gather is part of an all gather, and
+                    // the rank expects the send buffer to be populated with
+                    // the gather results. Thus, we read from the appropriate
+                    // offset.
+                    if (isInPlace) {
+                        memcpy(rankData.get() + r * sendOffset,
+                               sendBuffer + (sendRank * sendOffset),
+                               sendOffset);
+                    } else {
+                        memcpy(rankData.get() + r * sendOffset,
+                               sendBuffer,
+                               sendOffset);
+                    }
+
+                    continue;
+                }
+
+                recv(ranksForHost[thisHost].at(r),
+                     sendRank,
+                     rankData.get() + r * sendOffset,
+                     sendType,
+                     sendCount,
+                     nullptr,
+                     faabric::MPIMessage::GATHER);
+            }
+
+            send(sendRank,
+                 recvRank,
+                 rankData.get(),
+                 sendType,
+                 sendCount * ranksForHost[thisHost].size(),
+                 faabric::MPIMessage::GATHER);
+        } else {
+            // If we are the local leader but are co-located with the receiver
+            // of the gather, we just send our data for gathering.
+            // Note that we, again, make a distinction between sending with
+            // the in place flag or not.
+            if (isInPlace) {
```

Comment:
This code is very nested and with all the branching it's difficult to work out what's going on.

There are a few places it could be simplified where there's two `if`s on consecutive lines, or `else` and `if` on consecutive lines which can be merged, i.e.:

```cpp
if(a) { 

} else {
    if(b) {
    }
}
```

Can be:

```cpp
if(a) {

} else if(b) {

} 
```